pkg/io/uintn.go

```diff
@@ -0,0 +1,120 @@
+// Copyright 2012-2019 the u-root Authors. All rights reserved
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package io
+
+import (
+	"fmt"
+	"unsafe"
+)
+
+// UintN is a wrapper around uint types and provides a few io-related
+// functions.
+type UintN interface {
+	// Return size in bytes.
+	Size() int64
+
+	// Return string formatted in hex.
+	String() string
+
+	// Read from given address with native endianess.
+	read(addr unsafe.Pointer) error
+
+	// Write to given address with native endianess.
+	write(addr unsafe.Pointer) error
+}
+
+// Uint8 is a wrapper around uint8.
+type Uint8 uint8
+
+// Uint16 is a wrapper around uint16.
+type Uint16 uint16
+
+// Uint32 is a wrapper around uint32.
+type Uint32 uint32
+
+// Uint64 is a wrapper around uint64.
+type Uint64 uint64
+
+// Size of uint8 is 1.
+func (u *Uint8) Size() int64 {
+	return 1
+}
+
+// Size of uint16 is 2.
+func (u *Uint16) Size() int64 {
+	return 2
+}
+
+// Size of uint32 is 4.
+func (u *Uint32) Size() int64 {
+	return 4
+}
+
+// Size of uint64 is 8.
+func (u *Uint64) Size() int64 {
+	return 8
+}
+
+// String formats a uint8 in hex.
+func (u *Uint8) String() string {
+	return fmt.Sprintf("%#02x", uint8(*u))
+}
+
+// String formats a uint16 in hex.
+func (u *Uint16) String() string {
+	return fmt.Sprintf("%#04x", uint16(*u))
+}
+
+// String formats a uint32 in hex.
+func (u *Uint32) String() string {
+	return fmt.Sprintf("%#08x", uint32(*u))
+}
+
+// String formats a uint64 in hex.
+func (u *Uint64) String() string {
+	return fmt.Sprintf("%#016x", uint64(*u))
+}
+
+func (u *Uint8) read(addr unsafe.Pointer) error {
+	*u = Uint8(*(*uint8)(addr)) // TODO: rewrite in Go assembly for ARM
```

Comment:
Nice! I thought you cannot do it in Go.

Reply:
The assembly is doing the right thing for now, but future Go compilers might break this. I opened a bug: #1086